wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java

```diff
@@ -0,0 +1,446 @@
+
+		updateState(State.INPUT);
+	}
+
+	private void setText(final int resourceId, final String text)
+	{
+		TextView textView = getView(resourceId);
+		textView.setText(text);
+	}
+
+	private TextView getView(final int resourceId)
+	{
+		TextView textView = (TextView) findViewById(resourceId);
+		if (textView == null)
+		{
+			throw new IllegalArgumentException("Resource id " + resourceId + " not found.");
+		}
+		return textView;
+	}
+
+	private URL getUrl(final String pParam)
+	{
+		URL url;
+		try
+		{
+			url = new URL(pParam);
+		}
+		catch (MalformedURLException e)
+		{
+			Toast toast = new Toast(this);
+			toast.longToast("Not a proper destination URL:" + pParam);
+			finish();
+			return null;
+		}
+		return url;
+	}
+
+	@Override
+	public boolean onCreateOptionsMenu(final Menu menu)
+	{
+		getMenuInflater().inflate(R.menu.send_coins_activity_options, menu);
+
+		return super.onCreateOptionsMenu(menu);
+	}
+
+	@Override
+	public boolean onOptionsItemSelected(final MenuItem item)
+	{
+		switch (item.getItemId())
+		{
+			case android.R.id.home:
+				finish();
+				return true;
+
+			case R.id.send_coins_options_help:
+				HelpDialogFragment.page(getFragmentManager(), R.string.help_pop);
+				return true;
+		}
+
+		return super.onOptionsItemSelected(item);
+	}
+
+	public void sendPop(final View view)
+	{
+		View privateKeyBadPasswordView = findViewById(R.id.send_coins_private_key_bad_password);
+		privateKeyBadPasswordView.setVisibility(View.INVISIBLE);
+
+		SendPopTask sendPopTask = new SendPopTask();
+		sendPopTask.execute(popRequestURI);
+	}
+
+	public void resultMessage(final boolean success, final int resId, final Object... formatArgs)
+	{
+		TextView popMessage = getView(R.id.pop_message);
+
+		if (success)
+		{
+			Toast toast = new Toast(this);
+			toast.longToast(resId, formatArgs);
+			finish();
+		}
+		else
+		{
+			popMessage.setTextColor(getResources().getColor(R.color.fg_error));
+			popMessage.setVisibility(View.VISIBLE);
+			popMessage.setText(getString(resId, formatArgs));
+		}
+	}
+
+	public void resultMessage(final String message)
+	{
+		TextView popMessage = getView(R.id.pop_message);
+
+		popMessage.setTextColor(getResources().getColor(R.color.fg_error));
+		popMessage.setVisibility(View.VISIBLE);
+		popMessage.setText(message);
+	}
+
+	private void updateState(final State newState)
+	{
+		proveButton.setText(newState.getResId());
+		final boolean privateKeyPasswordViewVisible = getWalletApplication().getWallet().isEncrypted();
+		View pinGroup = findViewById(R.id.send_coins_private_key_password_group);
+		pinGroup.setVisibility(privateKeyPasswordViewVisible ? View.VISIBLE : View.GONE);
+		View privateKeyPasswordView = findViewById(R.id.send_coins_private_key_password);
+		privateKeyPasswordView.setEnabled(newState == State.INPUT);
+	}
+
+	private class Outcome
+	{
+		PopSender popSender;
+		Exception exception;
+	}
+
+	private class SendPopTask extends AsyncTask<PopRequestURI, State, Outcome>
```

Comment:
I generally avoid AsyncTasks in my apps. They're a pain to use (e.g. return values) and produce many spurious errors. Instead, I'm using a background handler, like in the original SendCoinsFragment.